src/libraries/System.Private.CoreLib/src/System/IO/FileStreamCompletionSource.Win32.cs

```diff
@@ -6,252 +6,262 @@
 using System.Runtime.InteropServices;
 using System.Threading;
 using System.Threading.Tasks;
+using Microsoft.Win32.SafeHandles;
 
 namespace System.IO
 {
-    internal sealed partial class LegacyFileStreamStrategy : FileStreamStrategy
+    // to avoid code duplicaiton of FileStreamCompletionSource for LegacyFileStreamStrategy and AsyncWindowsFileStreamStrategy
+    // we have created the following interface that is a common contract for both of them
+    internal interface IFileStreamCompletionSourceStrategy
```

Comment:
Why does this need an interface rather than just having the members be virtual on the Windows base type?  None of this code would be used for sync on Windows, so the base implementations could just assert and throw, right?

Reply:
It does not need to be an interface, but I consider it a lesser evil. First of all, I want to keep the dependencies between new and legacy strategies to a minimum.

Moreover, the inheritance depth is already big:

* `LegacysFileStreamStrategy : FileStreamStrategy : Stream` 
* `AsyncWindowsFileStreamStrategy : WindowsFileStreamStrategy : FileStreamStrategy : Stream`

By introducing one more base type we would have:

* `LegacysFileStreamStrategy : NewType : FileStreamStrategy : Stream` 
* `AsyncWindowsFileStreamStrategy : WindowsFileStreamStrategy : NewType : FileStreamStrategy : Stream`

And as you wrote, it would still require a virtual method dispatch so there would be no perf benefit?

Reply:
> And as you wrote, it would still require a virtual method dispatch so there would be no perf benefit?

Have you compared the cost of virtual dispatch to interface dispatch?

Reply:
> How are virtual methods any stronger a dependency?

I meant having a common base type with some logic.